akka-actor/src/main/scala-2.13+/akka/util/ByteString.scala

```diff
@@ -754,10 +815,29 @@ sealed abstract class ByteString
    */
   protected[ByteString] def toArray: Array[Byte] = toArray[Byte]
 
-  override def toArray[B >: Byte](implicit arg0: ClassTag[B]): Array[B] = iterator.toArray
+  final override def toArray[B >: Byte](implicit arg0: ClassTag[B]): Array[B] = {
+    // super uses byteiterator
+    val array = new Array[B](size)
+    copyToArray(array, 0, size)
+    array
+  }
+
+  final override def copyToArray[B >: Byte](xs: Array[B], start: Int): Int = {
+    // super uses byteiterator
+    copyToArray(xs, start, size.min(xs.size))
+  }
+
+  final override def copyToArray[B >: Byte](xs: Array[B], start: Int, len: Int): Int =
+    if (len > 0)
+      copyToArray(xs, 0, start, len)
+    else
+      len
 
-  override def copyToArray[B >: Byte](xs: Array[B], start: Int, len: Int): Int =
-    iterator.copyToArray(xs, start, len)
+  // optimized in all subclasses, avoiding usage of the iterator to save allocations/transformations
+  // same semantics as copyToArray(Array, start, len) wrt larger len than size of src and dest with one difference - if len or calculated negative it will throw
+  /** INTERNAL API */
+  @InternalApi
+  private[akka] def copyToArray[B >: Byte](dest: Array[B], srcPos: Int, destPos: Int, len: Int): Int
```

Comment:
This is basically needed to allow `ByteStrings` (the composed `ByteString` impl to allocate a single array and copy each sub-string into that at an offset.